### ISSUE TITLE
Auto completion and tagging not working

### ISSUE THREAD

**Issue by cyberphysicalrobotics** [MaskRay/ccls#239]
In main.cpp, When I jump to Node definition, it works. But when I tried company-lsp at "node.set", it doesn't work with the following error message.
Cannot complete at point
Did I missed some settings?

Files in directory:
ccls_project/src/main.cpp
ccls_project/src/node.cpp
ccls_project/include/ccls_project/node.hpp
ccls_project/CMakeLists.txt
ccls_project/compile_commands.json
ccls_project/.ccls
ccls_project/.ccls-root
ccls_project/.ccls-cache
```
main.cpp:
#include <iostream>
#include "ccls_project/node.hpp"

int main(int argc, char **argv) {

  Node node(0, 1, 2);
  std::cout << node.x() << node.theta() << std::endl;
  //node.set
  return 0;
}
```
```
in node.cpp:
#include "ccls_project/node.hpp"
void Node::set_f_value() { f_value_ = g_value_ + h_value_; }
```
```
#include <cmath>
#include <cstdint>
#include <limits>
#include <numeric>

#include "std_msgs/Time.h"

using std::numeric_limits;

enum Status
{
  not_expanded,
  open,
  closed
};

class Node
{
private:
  double x_ = 0;
  double y_ = 0;
  double theta_ = 0;
  double time_ = 0;

public:
  double f_value_ = 0;
  double g_value_ = numeric_limits<double>::max();
  double h_value_ = 0;
  double v_value_ = numeric_limits<double>::max();
  size_t hash_ = 0;
  Status status_ = Status::not_expanded;


  int prim_;
  Node* back_ptr_ = nullptr;

  Node() {}
  Node(double x, double y, double theta): x_(x), y_(y), theta_(theta) {}

  virtual ~Node() {}

  double x() const { return x_; }
  double y() const { return y_; }
  double theta() const { return theta_; }

  void set_f_value();
};
```
```
.ccls:
%compile_commands.json
%c -std=c11
%cpp -std=c++14
%c %cpp
%h %hpp
```

**Comment by MaskRay** [comment]
Please read the wiki page https://github.com/MaskRay/ccls/wiki/Project-Setup carefully

It has never suggested that `%c %cpp` can be applied to an empty flag. You can do `%c %cpp -DA` but not `%c %cpp`.

You need `compile_commands.json` if you use `%compile_commands.json`.

**Comment by MaskRay** [comment]
> 1.I don't understand -DA option. What does -DA mean in %c %cpp -DA?

`-DA` is like `#define A` in the source file.

> Because I'm not sure compile_commands.json file has right file informations for making database, If I want to add %cpp and %hpp to add file, can I use it like the following?

Don't do anything fancy if you don't understand the basics. Try the following example with an empty `.ccls` first:

`mkdir /tmp/c; cd /tmp/c; git init; touch .ccls; echo 'int a;' > a.cc` open a.cc

Then try `.ccls`:
```
%cpp -DA
```

`a.cc`:
```cpp
#ifdef A
int a;
#endif
```

> %cpp ./src/node.cpp
> %cpp ./src/main.cpp
> %hpp -I./include/ccls_project/node.hpp

No. If your `.ccls` contains `clang\n%cpp ./src/node.cpp`. When `a.cc` is opened, it is indexed as if `clang ./src/node.cpp a.cc` is used. There are two main files and it is definitely what you want.

**Comment by cyberphysicalrobotics** [comment]
Depending on where my header (hpp) files are, auto completion works or not. 
1.Case of header files in same directory with cpp files.
```
ccls_project/src/node.cpp
ccls_project/src/main.cpp
ccls_project/src/ccls_project/node.hpp
```
Picture 1. When header file is in same directory with cpp files (main.cpp, node.cpp), auto-completion works.
![image](https://user-images.githubusercontent.com/14055939/52167635-eb0d0f00-26d2-11e9-9608-694f3db46251.png)


2.Case of header files in same directory with cpp files.
```
ccls_project/src/node.cpp
ccls_project/src/main.cpp
ccls_project/include/ccls_project/node.hpp
```

Picture 2. When header file is in different directory with cpp files (main.cpp, node.cpp), auto-completion doesn't work
![image](https://user-images.githubusercontent.com/14055939/52167751-b732e900-26d4-11e9-86c9-7a421307fd82.png)

I've found that Node class (user-defined class) is shown as int type , not as in Node class type. I want to make sure how I can make tagging working right. In many case of big project, this cases happened a lot.

**Comment by MaskRay** [comment]
This is not a ccls issue. You don't understand how `-I` works. I suggest creating a shell scrip wrapper and pass `-v=1`.

> When header file is in same directory with cpp files (main.cpp, node.cpp), auto-completion works.

https://gcc.gnu.org/onlinedocs/gcc/Directory-Options.html

"For the quote form of the include directive, the directory of the current file is searched first."

> 2.In case of header files in same directory with cpp files, auto-completion doesn't work.

If you have `/tmp/ccls_project/include/ccls_project/node.hpp`, the current working directory is `/tmp/ccls_project` and you have `/tmp/ccls_project/.ccls` like the following:

```
clang
-Iinclude/ccls_project
```

You should use `#include "node.hpp"`, not `#include "ccls_project/node.hpp"`.

**Comment by cyberphysicalrobotics** [comment]
This is not a ccls issue. You don't understand how -I works. I suggest creating a shell scrip wrapper and pass -v=1.

> You mean pass -v=1 to ccls?

I followed your advice. Now auto-completion for header file works. 
In CMakeLists.txt:
```
include_directories(
 include
 include/ccls_project
  ${catkin_INCLUDE_DIRS}
)
```

**Comment by cyberphysicalrobotics** [comment]
When I move ccls_project (project) directory to other place which has no .git directory, ccls works well.
But ccls_project is under the directory which has .git, auto-completion for some place doesn't work. I want to make sure how the project directory is set. 
 
The directory structure is like the following:
```
~/catkin_ws/src/an_TA/.git
~/catkin_ws/src/an_TA/.ccls
~/catkin_ws/src/an_TA/.ccls-cache
~/catkin_ws/src/an_TA/Project5/ccls_project/.git
~/catkin_ws/src/an_TA/Project5/ccls_project/.ccls
~/catkin_ws/src/an_TA/Project5/ccls_project/.ccls-cache
```

1.The upper directory has .git
If the upper directory (an_TA) has .git, ccls-directory is generated. How can I make ccls_project project directory under an_TA while I need an_TA directory with .git?

**Comment by MaskRay** [comment]
For many language clients, when you open a file, it uses some indicator to find its associated project and the project is associated with one or more language servers.

`.ccls-root` is currently only supported by Emacs lsp-mode. ccls.el uses `ccls--suggest-project-root` (which locates `.ccls-root`) to override projectile (usually `.git`).

**Comment by cyberphysicalrobotics** [comment]
Even if I removed .git in ~/catkin_ws/src/an_TA and made .ccls-root in ccls_project directory. Still an_TA is recognized as project root like the following. Can I set the ccls--suggest-project-root? I tried it in M-x in emacs. It didn't work. I tried found it as variable (M-h v). It didn't show up. 

<img width="320" alt="image" src="https://user-images.githubusercontent.com/14055939/52187299-82608800-27e1-11e9-8359-c96633a9d604.png">
